libraries/chain/stake.cpp

```diff
@@ -6,56 +6,36 @@ using namespace int_arithmetic;
 int64_t recall_proxied_traversal(const cyberway::chaindb::ram_payer_info& ram, symbol_code token_code,
     const AgentIndex& agents_idx, const GrantIndex& grants_idx,
-    GrantItr& arg_grant_itr, int64_t pct, bool forced_erase = false
+    GrantItr& arg_grant_itr, int64_t share, bool forced_erase = false
 ) {
-
-    auto share = safe_pct(arg_grant_itr->share, pct);
     auto agent = get_agent(token_code, agents_idx, arg_grant_itr->agent_name);
 
     EOS_ASSERT(share >= 0, transaction_exception, "SYSTEM: share can't be negative");
     EOS_ASSERT(share <= agent->shares_sum, transaction_exception, "SYSTEM: incorrect share val");
     
     int64_t ret = 0;
-    auto grantor_funds = safe_prop(agent->get_total_funds(), arg_grant_itr->share, agent->shares_sum);
     
-    if (share > 0 && grantor_funds > 0) {
-        auto profit = std::max(grantor_funds - arg_grant_itr->granted, int64_t(0));
-        auto agent_fee_pct = safe_prop<int64_t>(std::min(agent->fee, arg_grant_itr->break_fee), profit, grantor_funds);
-        
-        auto share_fee = safe_pct(agent_fee_pct, share);
-        auto share_net = share - share_fee;
-        auto balance_ret = safe_prop(agent->balance, share_net, agent->shares_sum);
-        EOS_ASSERT(balance_ret <= agent->balance, transaction_exception, "SYSTEM: incorrect balance_ret val");
-
-        auto proxied_ret = 0;
-        auto grant_itr = grants_idx.lower_bound(grant_key(token_code, agent->account));
-        while ((grant_itr != grants_idx.end()) &&
-               (grant_itr->token_code   == token_code) &&
-               (grant_itr->grantor_name == agent->account))
-        {
-            auto to_recall = safe_prop(share_net, grant_itr->share, agent->shares_sum);
-            auto cur_pct = safe_prop<int64_t>(config::percent_100, share_net, agent->shares_sum);
-            proxied_ret += recall_proxied_traversal(ram, token_code, agents_idx, grants_idx, grant_itr, cur_pct);
-        }
-        EOS_ASSERT(proxied_ret <= agent->proxied, transaction_exception, "SYSTEM: incorrect proxied_ret val");
-
-        agents_idx.modify(*agent, [&](auto& a) {
-            a.set_balance(a.balance - balance_ret);
-            a.proxied -= proxied_ret;
-            a.own_share += share_fee;
-            a.shares_sum -= share_net;
-        });
-        
-        ret = balance_ret + proxied_ret;
+    auto balance_ret = safe_prop(agent->balance, share, agent->shares_sum);
+    EOS_ASSERT(balance_ret <= agent->balance, transaction_exception, "SYSTEM: incorrect balance_ret val");
+
+    auto proxied_ret = 0;
+    auto grant_itr = grants_idx.lower_bound(grant_key(token_code, agent->account));
+    while ((grant_itr != grants_idx.end()) && (grant_itr->token_code == token_code) && (grant_itr->grantor_name == agent->account)) {
+        auto cur_share = safe_prop(grant_itr->share, share, agent->shares_sum);
+        proxied_ret += recall_proxied_traversal(ram, token_code, agents_idx, grants_idx, grant_itr, cur_share);
```

Comment:
note: looks like we can `continue` here if `cur_share` becomes `0` (to stop useless recursion)